File: Source/Other/DatabaseAbstraction/SPPostgreSQLConnectionWrapper.m
//
//  SPPostgreSQLConnectionWrapper.m
//  Sequel Ace
//

#import "SPPostgreSQLConnectionWrapper.h"
#import "SPPostgreSQLResultWrapper.h"
#import "SPPostgreSQLStreamingResultWrapper.h"
#import "Frameworks/SPPostgreSQLFramework/Headers/sppostgresql_ffi.h"

@implementation SPPostgreSQLConnectionWrapper {
    SPPostgreSQLConnection *_conn;

    NSString *_host;
    NSString *_username;
    NSString *_password;
    NSString *_database;
    NSUInteger _port;
    BOOL _useSSL;
    NSUInteger _timeout;

    NSString *_lastErrorMessage;
    NSUInteger _lastErrorID;
    BOOL _lastQueryErrored;
    unsigned long long _rowsAffected;
    unsigned long long _lastInsertID;
}

- (instancetype)init {
    if (self = [super init]) {
        _conn = sp_postgresql_connection_create();
        _port = 5432;
        _useSSL = NO;
        _timeout = 30;
        _rowsAffected = 0;
        _lastInsertID = 0;
    }
    return self;
}

- (void)dealloc {
    if (_conn) {
        sp_postgresql_connection_destroy(_conn);
        _conn = NULL;
    }
}

+ (NSUInteger)defaultPort { return 5432; }

// ─── Configuration ───────────────────────────────────────────────────────────

- (void)setUsername:(NSString *)u  { _username = [u copy]; }
- (void)setPassword:(NSString *)p  { _password = [p copy]; }
- (void)setHost:(NSString *)h      { _host = [h copy]; }
- (void)setPort:(NSUInteger)port   { _port = port; }
- (void)setDatabase:(NSString *)db { _database = [db copy]; }
- (void)setTimeout:(NSUInteger)t   { _timeout = t; }
- (void)setUseSSL:(BOOL)ssl        { _useSSL = ssl; }
- (void)setDelegate:(id<SPDatabaseConnectionProxy>)delegate { /* PostgreSQL wrapper uses its own error handling. */ }

// ─── Connectivity ────────────────────────────────────────────────────────────

- (BOOL)connect {
    if (!_conn) { _conn = sp_postgresql_connection_create(); }

    const char *host     = _host     ? [_host UTF8String]     : "localhost";
    const char *user     = _username ? [_username UTF8String] : "postgres";
    const char *password = _password ? [_password UTF8String] : "";
    const char *database = _database ? [_database UTF8String] : "postgres";

    int result = sp_postgresql_connection_connect(_conn, host, (int)_port, user, password, database, _useSSL ? 1 : 0);

    if (!result) {
        [self _captureLastError];
        return NO;
    }
    _lastErrorMessage = nil;
    _lastErrorID = 0;
    return YES;
}

- (void)disconnect {
    if (_conn) sp_postgresql_connection_disconnect(_conn);
}

- (BOOL)isConnected {
    return _conn && sp_postgresql_connection_is_connected(_conn) != 0;
}

- (BOOL)isConnectedViaSSL { return _useSSL && [self isConnected]; }

// ─── Database selection ───────────────────────────────────────────────────────

- (BOOL)selectDatabase:(NSString *)aDatabase {
    // PostgreSQL does not support in-connection database switching; reconnect instead.
    if ([aDatabase isEqualToString:_database]) return YES;
    _database = [aDatabase copy];
    [self disconnect];
    return [self connect];
}

- (NSArray *)databases {
    if (!_conn) return @[];
    int count = 0;
    char **dbs = sp_postgresql_connection_list_databases(_conn, &count);
    if (!dbs || count == 0) return @[];

    NSMutableArray *arr = [NSMutableArray arrayWithCapacity:count];
    for (int i = 0; i < count; i++) {
        if (dbs[i]) [arr addObject:[NSString stringWithUTF8String:dbs[i]]];
    }
    sp_postgresql_free_string_array(dbs, count);
    return [arr copy];
}

// ─── Querying ────────────────────────────────────────────────────────────────

- (id<SPDatabaseResult>)queryString:(NSString *)query {
    if (!_conn || ![self isConnected]) {
        _lastErrorMessage = @"Not connected to PostgreSQL server";
        _lastErrorID = 1;
        _lastQueryErrored = YES;
        return nil;
    }

    NSTimeInterval start = [NSDate timeIntervalSinceReferenceDate];
    SPPostgreSQLResult *result = sp_postgresql_connection_execute_query(_conn, [query UTF8String]);
    double elapsed = [NSDate timeIntervalSinceReferenceDate] - start;

    if (!result) {
        [self _captureLastError];
        _lastQueryErrored = YES;
        return nil;
    }

    _lastQueryErrored = NO;
    _lastErrorMessage = nil;
    _lastErrorID = 0;
    _rowsAffected = sp_postgresql_result_affected_rows(result);
    _lastInsertID = 0; // PostgreSQL uses RETURNING instead of last-insert-id

    return [[SPPostgreSQLResultWrapper alloc] initWithResult:result queryTime:elapsed];
}

- (id)streamingQueryString:(NSString *)query {
    return [self streamingQueryString:query useLowMemoryBlockingStreaming:YES];
}

- (id)streamingQueryString:(NSString *)query useLowMemoryBlockingStreaming:(BOOL)fullStreaming {
    if (!_conn || ![self isConnected]) {
        _lastErrorMessage = @"Not connected to PostgreSQL server";
        _lastErrorID = 1;
        _lastQueryErrored = YES;
        return nil;
    }

    NSTimeInterval start = [NSDate timeIntervalSinceReferenceDate];
    int batchSize = fullStreaming ? 100 : 1000;
    SPPostgreSQLStreamingResult *result = sp_postgresql_connection_execute_streaming_query(
        _conn, [query UTF8String], batchSize);
    double elapsed = [NSDate timeIntervalSinceReferenceDate] - start;

    if (!result) {
        [self _captureLastError];
        _lastQueryErrored = YES;
        return nil;
    }

    _lastQueryErrored = NO;
    _lastErrorMessage = nil;
    _lastErrorID = 0;

    return [[SPPostgreSQLStreamingResultWrapper alloc] initWithStreamingResult:result queryTime:elapsed];
}

- (BOOL)queryErrored { return _lastQueryErrored; }
- (NSString *)lastErrorMessage { return _lastErrorMessage; }
- (NSUInteger)lastErrorID { return _lastErrorID; }
- (unsigned long long)rowsAffectedByLastQuery { return _rowsAffected; }
- (unsigned long long)lastInsertID { return _lastInsertID; }
- (void)cancelCurrentQuery { /* Best-effort: PostgreSQL async cancel would require a separate connection. */ }

// ─── Encoding ────────────────────────────────────────────────────────────────

- (NSString *)encoding { return @"utf8"; }
- (BOOL)setEncoding:(NSString *)encoding { return YES; /* PostgreSQL manages encoding per-connection. */ }
- (void)storeEncodingForRestoration { }
- (void)restoreStoredEncoding { }
- (NSString *)preferredUTF8Encoding { return @"UTF8"; }

// ─── Server info ─────────────────────────────────────────────────────────────

- (NSString *)serverVersionString {
    id<SPDatabaseResult> result = [self queryString:@"SELECT version()"];
    if (!result) return @"PostgreSQL (unknown version)";
    NSArray *row = [result getRowAsArray];
    return row.firstObject ?: @"PostgreSQL";
}

- (NSUInteger)_versionComponentAtIndex:(NSUInteger)idx {
    NSString *ver = [self serverVersionString];
    // "PostgreSQL 15.3 on ..." → extract "15.3"
    NSRegularExpression *re = [NSRegularExpression regularExpressionWithPattern:@"(\\d+)\\.(\\d+)(?:\\.(\\d+))?" options:0 error:nil];
    NSTextCheckingResult *match = [re firstMatchInString:ver options:0 range:NSMakeRange(0, ver.length)];
    if (!match || idx + 1 > match.numberOfRanges) return 0;
    NSRange range = [match rangeAtIndex:idx + 1];
    if (range.location == NSNotFound) return 0;
    return (NSUInteger)[[ver substringWithRange:range] integerValue];
}

- (NSUInteger)serverMajorVersion   { return [self _versionComponentAtIndex:0]; }
- (NSUInteger)serverMinorVersion   { return [self _versionComponentAtIndex:1]; }
- (NSUInteger)serverReleaseVersion { return [self _versionComponentAtIndex:2]; }

- (BOOL)serverVersionIsGreaterThanOrEqualTo:(NSUInteger)major
                               minorVersion:(NSUInteger)minor
                             releaseVersion:(NSUInteger)release {
    NSUInteger maj = self.serverMajorVersion;
    NSUInteger min = self.serverMinorVersion;
    NSUInteger rel = self.serverReleaseVersion;
    if (maj != major) return maj > major;
    if (min != minor) return min > minor;
    return rel >= release;
}

- (NSString *)getServerVariableValue:(NSString *)variable {
    NSString *query = [NSString stringWithFormat:@"SHOW %@", variable];
    id<SPDatabaseResult> result = [self queryString:query];
    if (!result) return nil;
    NSArray *row = [result getRowAsArray];
    return row.firstObject;
}

// ─── Collation helpers ────────────────────────────────────────────────────────

- (NSArray *)getCollationsForEncoding:(NSString *)encoding {
    // PostgreSQL uses collations from the operating system; not relevant for this UI.
    return @[];
}

// ─── Backend type ─────────────────────────────────────────────────────────────

- (BOOL)isPostgreSQL { return YES; }
- (NSString *)identifierQuoteCharacter { return @"\""; }

// ─── Private ──────────────────────────────────────────────────────────────────

- (void)_captureLastError {
    if (!_conn) return;
    char *err = sp_postgresql_connection_last_error(_conn);
    if (err) {
        _lastErrorMessage = [NSString stringWithUTF8String:err];
        sp_postgresql_free_string(err);
    } else {
        _lastErrorMessage = @"Unknown PostgreSQL error";
    }
    _lastErrorID = 2;
}

@end

File: Source/Other/DatabaseAbstraction/SPPostgreSQLConnectionWrapper.swift
//
//  SPPostgreSQLConnectionWrapper.swift
//  Sequel Ace
//

import Foundation

@objc(SPPostgreSQLConnectionWrapper)
final class SPPostgreSQLConnectionWrapper: NSObject, SPDatabaseConnection {
    private var conn: OpaquePointer?
    @objc var host: String?
    private var username: String?
    private var password: String?
    @objc var database: String?
    private var port: UInt = 5432
    private var useSSL = false

    private var storedLastErrorMessage: String?
    private var storedLastErrorID: UInt = 0
    private var lastQueryErrored = false
    private var rowsAffected: UInt64 = 0
    private var storedLastInsertID: UInt64 = 0

    @objc override init() {
        super.init()
        conn = sp_postgresql_connection_create()
    }

    deinit {
        if let conn {
            sp_postgresql_connection_destroy(conn)
            self.conn = nil
        }
    }

    @objc static var defaultPort: UInt { 5432 }

    func setUsername(_ username: String) { self.username = username }
    func setPassword(_ password: String) { self.password = password }
    func setPort(_ port: UInt) { self.port = port }
    func setTimeout(_ timeout: UInt) { /* stored by Rust connection on connect if needed */ }
    func setUseSSL(_ useSSL: Bool) { self.useSSL = useSSL }
    func setDelegate(_ delegate: SPDatabaseConnectionProxy?) { }

    func connect() -> Bool {
        if conn == nil { conn = sp_postgresql_connection_create() }
        guard let conn else { return false }

        let hostC = host ?? "localhost"
        let userC = username ?? "postgres"
        let passC = password ?? ""
        let dbC = database ?? "postgres"

        let result = hostC.withCString { hostPtr in
            userC.withCString { userPtr in
                passC.withCString { passPtr in
                    dbC.withCString { dbPtr in
                        sp_postgresql_connection_connect(
                            conn, hostPtr, Int32(port), userPtr, passPtr, dbPtr, useSSL ? 1 : 0
                        )
                    }
                }
            }
        }

        if result == 0 {
            captureLastError()
            return false
        }
        storedLastErrorMessage = nil
        storedLastErrorID = 0
        return true
    }

    func disconnect() {
        if let conn { sp_postgresql_connection_disconnect(conn) }
    }

    func isConnected() -> Bool {
        guard let conn else { return false }
        return sp_postgresql_connection_is_connected(conn) != 0
    }

    func isConnectedViaSSL() -> Bool { useSSL && isConnected() }

    func selectDatabase(_ aDatabase: String) -> Bool {
        if aDatabase == database { return true }
        database = aDatabase
        disconnect()
        return connect()
    }

    func databases() -> [String] {
        guard let conn else { return [] }
        var count: Int32 = 0
        guard let dbs = sp_postgresql_connection_list_databases(conn, &count), count > 0 else {
            return []
        }
        defer { sp_postgresql_free_string_array(dbs, count) }

        var names = [String]()
        for i in 0..<count {
            if let ptr = dbs[Int(i)] {
                names.append(String(cString: ptr))
            }
        }
        return names
    }

    func queryString(_ query: String) -> SPDatabaseResult? {
        guard let conn, isConnected() else {
            storedLastErrorMessage = "Not connected to PostgreSQL server"
            storedLastErrorID = 1
            lastQueryErrored = true
            return nil
        }

        let start = Date.timeIntervalSinceReferenceDate
        let result: OpaquePointer? = query.withCString { queryPtr in
            sp_postgresql_connection_execute_query(conn, queryPtr)
        }
        let elapsed = Date.timeIntervalSinceReferenceDate - start

        guard let result else {
            captureLastError()
            lastQueryErrored = true
            return nil
        }

        lastQueryErrored = false
        storedLastErrorMessage = nil
        storedLastErrorID = 0
        rowsAffected = sp_postgresql_result_affected_rows(result)
        storedLastInsertID = 0

        return SPPostgreSQLResultWrapper(result: result, queryTime: elapsed)
    }

    func streamingQueryString(_ query: String) -> Any? {
        streamingQueryString(query, useLowMemoryBlockingStreaming: true)
    }

    func streamingQueryString(_ query: String, useLowMemoryBlockingStreaming fullStreaming: Bool) -> Any? {
        guard let conn, isConnected() else {
            storedLastErrorMessage = "Not connected to PostgreSQL server"
            storedLastErrorID = 1
            lastQueryErrored = true
            return nil
        }

        let start = Date.timeIntervalSinceReferenceDate
        let batchSize = fullStreaming ? 100 : 1000
        let result: OpaquePointer? = query.withCString { queryPtr in
            sp_postgresql_connection_execute_streaming_query(conn, queryPtr, Int32(batchSize))
        }
        let elapsed = Date.timeIntervalSinceReferenceDate - start

        guard let result else {
            captureLastError()
            lastQueryErrored = true
            return nil
        }

        lastQueryErrored = false
        storedLastErrorMessage = nil
        storedLastErrorID = 0

        return SPPostgreSQLStreamingResultWrapper(streamingResult: result, queryTime: elapsed)
    }

    func queryErrored() -> Bool { lastQueryErrored }
    func lastErrorMessage() -> String? { storedLastErrorMessage }
    func lastErrorID() -> UInt { storedLastErrorID }
    func rowsAffectedByLastQuery() -> UInt64 { rowsAffected }
    func lastInsertID() -> UInt64 { storedLastInsertID }
    func cancelCurrentQuery() { }

    func encoding() -> String? { "utf8" }
    func setEncoding(_ encoding: String) -> Bool { true }
    func storeEncodingForRestoration() { }
    func restoreStoredEncoding() { }
    func preferredUTF8Encoding() -> String { "UTF8" }

    func serverVersionString() -> String? {
        guard let result = queryString("SELECT version()"),
              let row = result.getRowAsArray(),
              let version = row.first as? String else {
            return "PostgreSQL (unknown version)"
        }
        return version
    }

    private func versionComponent(at index: Int) -> UInt {
        guard let ver = serverVersionString() else { return 0 }
        let pattern = "(\\d+)\\.(\\d+)(?:\\.(\\d+))?"
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: ver, range: NSRange(ver.startIndex..., in: ver)),
              index + 1 < match.numberOfRanges else { return 0 }
        let range = match.range(at: index + 1)
        guard range.location != NSNotFound, let swiftRange = Range(range, in: ver) else { return 0 }
        return UInt(ver[swiftRange]) ?? 0
    }

    func serverMajorVersion() -> UInt { versionComponent(at: 0) }
    func serverMinorVersion() -> UInt { versionComponent(at: 1) }
    func serverReleaseVersion() -> UInt { versionComponent(at: 2) }

    func serverVersionIsGreaterThanOrEqual(to major: UInt, minorVersion minor: UInt, releaseVersion release: UInt) -> Bool {
        let maj = serverMajorVersion()
        let min = serverMinorVersion()
        let rel = serverReleaseVersion()
        if maj != major { return maj > major }
        if min != minor { return min > minor }
        return rel >= release
    }

    func getServerVariableValue(_ variable: String) -> String? {
        guard let result = queryString("SHOW \(variable)"),
              let row = result.getRowAsArray(),
              let value = row.first as? String else { return nil }
        return value
    }

    func getCollationsForEncoding(_ encoding: String) -> [Any] { [] }

    func isPostgreSQL() -> Bool { true }
    func identifierQuoteCharacter() -> String { "\"" }

    @objc func escapeAndQuoteString(_ theString: String) -> String? {
        let escaped = theString
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "'", with: "''")
        return "'\(escaped)'"
    }

    @objc var lastQueryWasCancelled: Bool { false }

    private func captureLastError() {
        guard let conn else { return }
        if let errPtr = sp_postgresql_connection_last_error(conn) {
            storedLastErrorMessage = String(cString: errPtr)
            sp_postgresql_free_string(errPtr)
        } else {
            storedLastErrorMessage = "Unknown PostgreSQL error"
        }
        storedLastErrorID = 2
    }
}
